### PR TITLE
Deck Editor layout

### LINKED ISSUE #1329: "Both" deck editor display mode has a very small image
Card images when using the "Both" display mode on the deck editor are far too small and are wasting lots of space. I want to have both the card text and the image visible if possible.
screenshot 2015-08-04 12 29 00

I can understand that if the window is reduced in size then having the small image makes sense but when the program is full screened it seems very un-intuitive.

### PR DESCRIPTION
**Pull request by marcofernandezheras**
Fix #1329

Deck editor layout based on dock widgets

docks_1

docks_2

docks_3

### CODE REVIEW

File: cockatrice/src/settingscache.cpp [hunk 1]
```diff
@@ -1,8 +1,12 @@
 #include "settingscache.h"
 #include <QSettings>
 
+#include <QDebug>
+
+
+
```
Comment (ZeldaZach) [inline cockatrice/src/settingscache.cpp]: Do we need QDebug?
Also, extra line spaces

File: cockatrice/src/settingscache.cpp [hunk 2]
```diff
@@ -1,8 +1,12 @@
...
 SettingsCache::SettingsCache()
-{
+{
```
Comment (ZeldaZach) [inline cockatrice/src/settingscache.cpp]: Extra spaceeeeee

File: cockatrice/src/tab_deck_editor.cpp [hunk 3]
```diff
@@ -267,19 +302,74 @@ TabDeckEditor::TabDeckEditor(TabSupervisor *_tabSupervisor, QWidget *parent)
...
+
+void TabDeckEditor::createCentralFrame()
+{
+    searchEdit = new SearchLineEdit;
+    searchEdit->setObjectName("searchEdit");
+#if QT_VERSION >= 0x050300
+    searchEdit->addAction(QIcon(":/resources/icon_search_black.svg"), QLineEdit::LeadingPosition);
+#endif
+    searchEdit->setObjectName("searchEdit");
+
```
Comment (ZeldaZach) [inline cockatrice/src/tab_deck_editor.cpp]: This is defined 4 lines up already. Do we need to call this twice?

File: cockatrice/src/tab_deck_editor.cpp [hunk 4]
```diff
@@ -304,15 +399,82 @@ TabDeckEditor::TabDeckEditor(TabSupervisor *_tabSupervisor, QWidget *parent)
...
+    this->installEventFilter(this);    
+
+    retranslateUi();
     QTimer::singleShot(0, this, SLOT(checkFirstRunDetected()));
+    QTimer::singleShot(0, this, SLOT(loadLayout()));
 }
 
 TabDeckEditor::~TabDeckEditor()
-{
+{
```
Comment (ZeldaZach) [inline cockatrice/src/tab_deck_editor.cpp]: Spaceeeeeee

File: cockatrice/src/tab_deck_editor.h [hunk 5]
```diff
@@ -99,14 +107,25 @@ class TabDeckEditor : public Tab {
...
+    QMainWindow *MainWindow;
+    QVBoxLayout *centralFrame;
+    QHBoxLayout *searchLayout;
+    QPushButton *btnFilter;
+    QPushButton *btnDeck;
+    QPushButton *btnCard;
+    QDockWidget *cardInfoDock;
+    QDockWidget *deckDock;
+    QDockWidget *filterDock;
+    QWidget * centralWidget;
```
Comment (ZeldaZach) [inline cockatrice/src/tab_deck_editor.h]: There a space between the star and the var name.

**Comment by woogerboy21** [comment]
:+1:  This is awesome.  Now if the game tab worked the same way that would be sweet.
And why do I find it amusing that one of the commits is to pass @ZeldaZach code style inspection? :laughing:

File: cockatrice/src/cardinfopicture.cpp [hunk 6]
```diff
@@ -8,14 +8,12 @@
...
 #include "main.h"
 
-CardInfoPicture::CardInfoPicture(int width, QWidget *parent)
+CardInfoPicture::CardInfoPicture(QWidget *parent)
     : QWidget(parent),
     info(0),
     pixmapDirty(true)
 {
-    setFixedWidth(width);
     setMinimumHeight(100);
```
Comment (ZeldaZach) [inline cockatrice/src/cardinfopicture.cpp]: Do we still need a minimum height call?

File: cockatrice/src/settingscache.cpp [hunk 7]
```diff
@@ -462,6 +465,18 @@ QStringList SettingsCache::getCountries() const
     return countries;
 }
 
+void SettingsCache::setDeckEditorLayoutState(const QByteArray &value)  {
```
Comment (ZeldaZach) [inline cockatrice/src/settingscache.cpp]: The `{` should be on a new line to stay consistent.

File: cockatrice/src/settingscache.cpp [hunk 8]
```diff
@@ -462,6 +465,18 @@ QStringList SettingsCache::getCountries() const
...
+    deckEditorLayoutState = value;
+    QSettings layout_settings("deckLayout.ini" , QSettings::IniFormat);
+    layout_settings.setValue("layouts/deckEditor_state",value);
+}
+
+void SettingsCache::setDeckEditorGeometry(const QByteArray &value)  {
```
Comment (ZeldaZach) [inline cockatrice/src/settingscache.cpp]: The `{` should be on a new line to stay consistent. (here too)

File: cockatrice/src/settingscache.h [hunk 9]
```diff
@@ -170,8 +171,15 @@ class SettingsCache : public QObject {
     bool getSpectatorsCanTalk() const { return spectatorsCanTalk; }
     bool getSpectatorsCanSeeEverything() const { return spectatorsCanSeeEverything; }
     int getKeepAlive() const { return keepalive; }
+
+    QByteArray getDeckEditorLayoutState() const  { return deckEditorLayoutState; }
```
Comment (ZeldaZach) [inline cockatrice/src/settingscache.h]: extra space after `const`

File: cockatrice/src/settingscache.h [hunk 10]
```diff
@@ -170,8 +171,15 @@ class SettingsCache : public QObject {
...
+    void setDeckEditorLayoutState(const QByteArray &value);
+    QByteArray getDeckEditorGeometry() const  { return deckEditorGeometry;  }
```
Comment (ZeldaZach) [inline cockatrice/src/settingscache.h]: extra space after `const`  againnn

File: cockatrice/src/tab_deck_editor.cpp [hunk 11]
```diff
@@ -267,19 +302,73 @@ TabDeckEditor::TabDeckEditor(TabSupervisor *_tabSupervisor, QWidget *parent)
     deckMenu->addSeparator();
     deckMenu->addAction(aAnalyzeDeck);
     deckMenu->addSeparator();
+    deckMenu->addAction(aResetLayout);
```
Comment (ZeldaZach) [inline cockatrice/src/tab_deck_editor.cpp]: Can you move this in between `aLoadDeck` and `aAnalyzeDeck`?

File: cockatrice/src/cardinfopicture.cpp (hunk 6 above)
Comment (woogerboy21) [inline cockatrice/src/cardinfopicture.cpp]: This seems to fix the issue with the small card image on the "both" tab.

Comment (ZeldaZach) [inline cockatrice/src/cardinfopicture.cpp]: Gotcha, then this is fine.

**Comment by ZeldaZach** [comment]
When you attempt to reset the layout, all elements are returned to their proper locations, but the card database list doesn't resize if you made it larger/smaller 

(These were taken after I resized it and then hit the "Reset layout" button)
screenshot 2015-08-06 23 08 25
screenshot 2015-08-06 23 09 03

**Comment by Daenyth** [comment]
We should add an auto style linter to travis.

File: cockatrice/src/settingscache.h [hunk 12]
```diff
@@ -170,8 +175,20 @@ class SettingsCache : public QObject {
...
+    void setDeckEditorGeometry(const QByteArray &value);
+    QSize getDeckEditorCardSize() const { return deckEditorCardSize; }
+    void setDeckEditorCardSize(const QSize &value);
+    QSize getDeckEditorDeckSize() const { return deckEditorDeckSize; }
+    void setDeckEditorDeckSize(const QSize &value);
+    QSize getDeckEditorFilterSize() const { return deckEditorFilterSize; }
+    void setDeckEditorFilterSize(const QSize &value);
     void setClientID(QString clientID);
-    QString getClientID() { return clientID; }
+    QString getClientID() { return clientID; }
```
Comment (ZeldaZach) [inline cockatrice/src/settingscache.h]: I spy extra space ;p

**Comment by ZeldaZach** [comment]
This seems good to go to me, just a `git rebase` is needed first :smile: :shipit:

**Comment by woogerboy21** [comment]
Why is my face under the last commit?

File: cockatrice/src/tab_deck_editor.cpp [hunk 13]
```diff
@@ -201,26 +140,119 @@ TabDeckEditor::TabDeckEditor(TabSupervisor *_tabSupervisor, QWidget *parent)
...
+    cardInfoDock->setMinimumSize(QSize(200, 41));
+    cardInfoDock->setAllowedAreas(Qt::LeftDockWidgetArea|Qt::RightDockWidgetArea);
+    cardInfoDock->setFeatures(QDockWidget::DockWidgetClosable|QDockWidget::DockWidgetFloatable|QDockWidget::DockWidgetMovable);
+    QWidget *cardInfoDockContents = new QWidget();
+    cardInfoDockContents->setObjectName("cardInfoDockContents");
+    cardInfoDockContents->setLayout(cardInfoFrame);
+    cardInfoDock->setWidget(cardInfoDockContents);
+
+    connect(btnCard,SIGNAL(toggled(bool)),cardInfoDock,SLOT(setVisible(bool)));
+    cardInfoDock->installEventFilter(this);
```
Comment (ZeldaZach) [inline cockatrice/src/tab_deck_editor.cpp]: Spaceeeeeee

File: cockatrice/src/tab_deck_editor.cpp [hunk 14]
```diff
@@ -201,26 +140,119 @@ TabDeckEditor::TabDeckEditor(TabSupervisor *_tabSupervisor, QWidget *parent)
...
+    filterFrame->addWidget(filterBox);
+
+    filterDock = new QDockWidget(MainWindow);
+    filterDock->setObjectName("filterDock");
+
+    filterDock->setFeatures(QDockWidget::DockWidgetClosable|QDockWidget::DockWidgetFloatable|QDockWidget::DockWidgetMovable);
+    QWidget *filterDockContents = new QWidget(MainWindow);
+    filterDockContents->setObjectName("filterDockContents");
+    filterDockContents->setLayout(filterFrame);
+    filterDock->setWidget(filterDockContents);
```
Comment (ZeldaZach) [inline cockatrice/src/tab_deck_editor.cpp]: More space

File: cockatrice/src/tab_deck_editor.cpp [hunk 15]
```diff
@@ -267,19 +302,74 @@ TabDeckEditor::TabDeckEditor(TabSupervisor *_tabSupervisor, QWidget *parent)
...
+    searchEdit->installEventFilter(&searchKeySignals);
+    searchKeySignals.setObjectName("searchKeySignals");
+    connect(searchEdit, SIGNAL(textChanged(const QString &)), this, SLOT(updateSearch(const QString &)));
+    connect(&searchKeySignals, SIGNAL(onEnter()), this, SLOT(actAddCard()));
+    connect(&searchKeySignals, SIGNAL(onCtrlAltEqual()), this, SLOT(actAddCard()));
+    connect(&searchKeySignals, SIGNAL(onCtrlAltRBracket()), this, SLOT(actAddCardToSideboard()));
+    connect(&searchKeySignals, SIGNAL(onCtrlAltMinus()), this, SLOT(actDecrementCard()));
+    connect(&searchKeySignals, SIGNAL(onCtrlAltLBracket()), this, SLOT(actDecrementCardFromSideboard()));
+    connect(&searchKeySignals, SIGNAL(onCtrlAltEnter()), this, SLOT(actAddCardToSideboard()));
+    connect(&searchKeySignals, SIGNAL(onCtrlEnter()), this, SLOT(actAddCardToSideboard()));
```
Comment (ZeldaZach) [inline cockatrice/src/tab_deck_editor.cpp]: Spacearino

File: cockatrice/src/tab_deck_editor.cpp [hunk 16]
```diff
@@ -304,11 +399,108 @@ TabDeckEditor::TabDeckEditor(TabSupervisor *_tabSupervisor, QWidget *parent)
...
+    centralWidget = new QWidget(MainWindow);
+    centralWidget->setObjectName("centralWidget");
+    centralWidget->setLayout(centralFrame);
+    MainWindow->setCentralWidget(centralWidget);
+    MainWindow->setDockOptions(QMainWindow::AnimatedDocks|QMainWindow::AllowNestedDocks|QMainWindow::AllowTabbedDocks);
+
+    QHBoxLayout *mainLayout = new QHBoxLayout;
+    mainLayout->setObjectName("mainLayout");
+    mainLayout->addWidget(MainWindow);
+    setLayout(mainLayout);
```
Comment (ZeldaZach) [inline cockatrice/src/tab_deck_editor.cpp]: Spaceeee

File: cockatrice/src/tab_deck_editor.cpp [hunk 17]
```diff
@@ -304,11 +399,108 @@ TabDeckEditor::TabDeckEditor(TabSupervisor *_tabSupervisor, QWidget *parent)
...
+
+    createCentralFrame();
+
+    createDeckDock();
+    createCardInfoDock();
+    createFiltersDock();
+
+    restartLayout();
+
+    this->installEventFilter(this);
```
Comment (ZeldaZach) [inline cockatrice/src/tab_deck_editor.cpp]: Spaceee

File: cockatrice/src/tab_deck_editor.cpp [hunk 18]
```diff
@@ -737,6 +936,27 @@ void TabDeckEditor::setPriceTagFeatureEnabled(int /* enabled */)
...
+bool TabDeckEditor::eventFilter(QObject * o, QEvent * e)
+{
+    if(e->type() == QEvent::Close)
+    {
+        if(o == cardInfoDock)
+            btnCard->setChecked(false);
+        else if(o == deckDock)
+            btnDeck->setChecked(false);
+        else if(o == filterDock)
+            btnFilter->setChecked(false);
```
Comment (ZeldaZach) [inline cockatrice/src/tab_deck_editor.cpp]: Lot o space

File: cockatrice/src/tab_deck_editor.cpp [hunk 19]
```diff
@@ -737,6 +936,27 @@ void TabDeckEditor::setPriceTagFeatureEnabled(int /* enabled */)
...
+{
+    if(e->type() == QEvent::Close)
+    {
+        if(o == cardInfoDock)
+            btnCard->setChecked(false);
+        else if(o == deckDock)
+            btnDeck->setChecked(false);
+        else if(o == filterDock)
+            btnFilter->setChecked(false);            
+    }
```
Comment (ZeldaZach) [inline cockatrice/src/tab_deck_editor.cpp]: Spacee

File: cockatrice/src/tab_game.cpp [hunk 20]
```diff
@@ -282,7 +282,7 @@ TabGame::TabGame(TabSupervisor *_tabSupervisor, GameReplay *_replay)
     gameView = new GameView(scene);
     gameView->hide();
```
Comment (ZeldaZach) [inline cockatrice/src/tab_game.cpp]: Space on blank line

File: cockatrice/src/tab_game.cpp [hunk 21]
```diff
@@ -282,7 +282,7 @@ TabGame::TabGame(TabSupervisor *_tabSupervisor, GameReplay *_replay)
...
     
-    cardInfo = new CardFrame(250, 372);
+    cardInfo = new CardFrame();
     playerListWidget = new PlayerListWidget(0, 0, this);
     playerListWidget->setFocusPolicy(Qt::NoFocus);
```
Comment (ZeldaZach) [inline cockatrice/src/tab_game.cpp]: Space on blank line

File: cockatrice/src/tab_game.cpp [hunk 22]
```diff
@@ -396,7 +396,7 @@ TabGame::TabGame(TabSupervisor *_tabSupervisor, QList<AbstractClient *> &_client
     gameView = new GameView(scene);
     gameView->hide();
```
Comment (ZeldaZach) [inline cockatrice/src/tab_game.cpp]: Space on blank line

File: cockatrice/src/tab_deck_editor.cpp [hunk 23]
```diff
@@ -201,26 +140,119 @@ TabDeckEditor::TabDeckEditor(TabSupervisor *_tabSupervisor, QWidget *parent)
...
+void TabDeckEditor::createCardInfoDock()
+{
+    cardInfo = new CardFrame();
+    cardInfo->setObjectName("cardInfo");
+    QVBoxLayout *cardInfoFrame = new QVBoxLayout;
+    cardInfoFrame->setObjectName("cardInfoFrame");
+    cardInfoFrame->addWidget(cardInfo);
+
+    cardInfoDock = new QDockWidget(MainWindow);
+    cardInfoDock->setObjectName("cardInfoDock");
```
Comment (ZeldaZach) [inline cockatrice/src/tab_deck_editor.cpp]: Spaceee

**Comment by ZeldaZach** [comment]
Final chance for reviews. I'm shipping this later tonight :shipit:

**Pull request merged**